sound/soc/sof/intel/hda-loader.c

```diff
@@ -373,26 +376,147 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/* Code loader helper APIs */
+static int cl_skl_cldma_setup_bdle(struct snd_sof_dev *sdev,
+				    struct snd_dma_buffer *dmab_data,
+				    u32 **bdlp, int size, int with_ioc)
+{
+	u32 *bdl = *bdlp;
+	int frags = 0;
+
+	while (size > 0) {
+		phys_addr_t addr = virt_to_phys(dmab_data->area +
+						(frags * size));
+
+		bdl[0] = cpu_to_le32(lower_32_bits(addr));
+		bdl[1] = cpu_to_le32(upper_32_bits(addr));
+
+		bdl[2] = cpu_to_le32(size);
+
+		size -= size;
+		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
+
+		bdl += 4;
+		frags++;
+	}
+
+	return frags;
+}
+
+static void cl_skl_cldma_stream_run(struct snd_sof_dev *sdev, bool enable)
+{
+	unsigned char val;
+	int timeout;
+
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SKL_ADSP_REG_CL_SD_CTL,
+				CL_SD_CTL_RUN_MASK, CL_SD_CTL_RUN(enable));
+
+	udelay(3);
+	timeout = 300;
+	do {
+		/* waiting for hardware to report the stream Run bit set */
+		val = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+				       SKL_ADSP_REG_CL_SD_CTL) &
+				       CL_SD_CTL_RUN_MASK;
+		if (enable && val)
+			break;
+		else if (!enable && !val)
+			break;
+		udelay(3);
+	} while (--timeout);
+
+	if (timeout == 0)
+		dev_err(sdev->dev, "Failed to set Run bit=%d enable=%d\n",
+			val, enable);
+}
+
+/*
+ * Setup controller
+ * Configure the registers to update the dma buffer address and
+ * enable interrupts.
+ * Note: Using the channel 1 for transfer
+ */
+static void cl_skl_cldma_setup_controller(struct snd_sof_dev *sdev,
+		struct snd_dma_buffer *dmab_bdl, unsigned int max_size,
+		u32 count)
+{
+	/* make sure Run bit is cleared before setting stream register */
```

Comment:
ok, so I guess this is where we differ from APL, if so is it all the register write below or only some ?